### PR TITLE
Dispose on deinit

### PR DESCRIPTION
**Pull request by tim-chamberlin**
RFC: https://docs.google.com/document/d/1gxfNxSg5yNO8-O-zkDvXTCXdNLJbne014ahnM4zzA6A/edit#

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,14 +44,24 @@ let observable = Observable<thing>()
...
-let disposer = Disposer()
+extension MyViewController {
+  button.tap.subscribe(onNext: { [weak self] in
+    self?.navigationController.push(newVc)
+  }).add(to: disposer)
+}
 ```
+Without specifying a `[weak self]` capture list in a scenario like this, a retain cycle is created between the subscriber and the view controller. In this example, without the capture list, the view controller will not be deallocated as expected, causing its disposer object to stay in memory as well. Since the `Disposer` removes its referenced subscribers when it is deinitialized, these subscribers will stay in memory as well.
+
+See [https://docs.swift.org/swift-book/LanguageGuide/AutomaticReferenceCounting.html](https://docs.swift.org/swift-book/LanguageGuide/AutomaticReferenceCounting.html) for more details on memory management in Swift.
```
Comment (luispadron) [inline README.md:64]: Awesome 👍🏼

File: Snail/Disposer.swift [hunk 2]
```diff
@@ -8,15 +8,24 @@ public protocol DisposableType {
 
 public class Disposer {
     private(set) public var disposables: [DisposableType] = []
+    private let semaphore = DispatchSemaphore(value: 1)
```
Comment (luispadron) [inline Snail/Disposer.swift:11]: Instead of a semaphore, I'd go with a read / write queue. This way we ensure writes to the `disposables` array are safe without blocking the entire thread.

Take a look at: https://github.com/UrbanCompass/Snail/blob/main/Snail/Observable.swift#L10 for how we do this

File: Snail/Observable.swift [hunk 3]
```diff
@@ -6,7 +6,7 @@ import Dispatch
 public class Observable<T>: ObservableType {
     private var isStopped: Int32 = 0
     private var stoppedEvent: Event<T>?
-    private var subscribers: [Subscriber<T>] = []
+    private(set) var subscribers: [Subscriber<T>] = []
```
Comment (luispadron) [inline Snail/Observable.swift:9]: Is this for testing purposes?

File: Snail/Observable.swift [hunk 4]
```diff
@@ -21,7 +21,7 @@ public class Observable<T>: ObservableType {
         }
     }
 
-    @discardableResult public func subscribe(queue: DispatchQueue? = nil, onNext: ((T) -> Void)? = nil, onError: ((Error) -> Void)? = nil, onDone: (() -> Void)? = nil) -> Subscriber<T> {
```
Comment (luispadron) [inline Snail/Observable.swift:24]: heck ya!!

File: Snail/Observable.swift [hunk 5]
```diff
@@ -61,7 +61,7 @@ public class Observable<T>: ObservableType {
 
     public func on(_ queue: DispatchQueue) -> Observable<T> {
         let observable = Observable<T>()
-        subscribe(queue: queue,
+        _ = subscribe(queue: queue,
                   onNext: { observable.on(.next($0)) },
                   onError: { observable.on(.error($0)) },
                   onDone: { observable.on(.done) })
```
Comment (luispadron) [inline Snail/Observable.swift:67]: I think we should add this to the subscribers array right? This way we remove this subscription when `deinit`

File: Snail/Observable.swift [hunk 6]
```diff
@@ -89,7 +89,7 @@ public class Observable<T>: ObservableType {
     public func map<U>(_ transform: @escaping (T) -> U) -> Observable<U> {
         let transformed = Observable<U>()
 
-        subscribe(
+        _ = subscribe(
```
Comment (luispadron) [inline Snail/Observable.swift:92]: I think we should add this to the subscribers array right? This way we remove this subscription when `deinit`

File: Snail/Observable.swift [hunk 7]
```diff
@@ -107,7 +107,7 @@ public class Observable<T>: ObservableType {
     public func flatMap<U>( _ transform: @escaping (T) -> Observable<U>) -> Observable<U> {
         let flatMapped = Observable<U>()
 
-        subscribe(
+        _ = subscribe(
```
Comment (luispadron) [inline Snail/Observable.swift:110]: I think we should add this to the subscribers array right? This way we remove this subscription when `deinit`

File: Snail/Observable.swift [hunk 8]
```diff
@@ -126,7 +126,7 @@ public class Observable<T>: ObservableType {
     public func filter(_ isIncluded: @escaping (T) -> Bool) -> Observable<T> {
         let filtered = Observable<T>()
 
-        subscribe(
+        _ = subscribe(
```
Comment (luispadron) [inline Snail/Observable.swift:129]: I think we should add this to the subscriber's array right? This way we remove this subscription when `deinit`

File: Snail/Observable.swift [hunk 9]
```diff
@@ -155,7 +155,7 @@ public class Observable<T>: ObservableType {
 
         let semaphore = DispatchSemaphore(value: 0)
 
-        subscribe(onNext: { value in
+        _ = subscribe(onNext: { value in
```
Comment (luispadron) [inline Snail/Observable.swift:158]: I think we should add this to the subscribers array right? This way we remove this subscription when `deinit`

File: Snail/Observable.swift [hunk 10]
```diff
@@ -188,15 +188,15 @@ public class Observable<T>: ObservableType {
         scheduler.start()
 
         var next: T?
-        scheduler.observable.subscribe(onNext: {
+        _ = scheduler.observable.subscribe(onNext: {
```
Comment (luispadron) [inline Snail/Observable.swift:191]: I think we should add this to the subscribers array right? This way we remove this subscription when `deinit`

File: SnailTests/ObservableTests.swift [hunk 11]
```diff
@@ -924,4 +925,35 @@ class ObservableTests: XCTestCase {
             subject.on(.next("main - async"))
         }
     }
+
+    func testDisposerRemovesSubscriptionFromObserverOnDeinit() {
```
Comment (luispadron) [inline SnailTests/ObservableTests.swift:929]: Does this test fail if we undo the changes this PR makes? If so ❤️ , if not lets come up with a use case here!

We should also consider doing this same test for the operators: `.map`, `.merge`, etc.

**Review by luispadron — CHANGES_REQUESTED** [review]
Overall great direction, this is going to be great 👍🏼

File: Snail/Observable.swift (hunk 3 above)
Comment (tim-chamberlin) [reply Snail/Observable.swift:9]: yeah I needed to change the getter to internal to run your unit test.

File: Snail/Disposer.swift (hunk 2 above)
Comment (tim-chamberlin) [reply Snail/Disposer.swift:11]: yeah that makes sense. good call

File: Snail/Observable.swift (hunk 6 above)
Comment (tim-chamberlin) [reply Snail/Observable.swift:92]: `.subscribe()` adds the subscription to the subscribers array.

**Review by luispadron — DISMISSED** [review]

File: Snail/Variable.swift [hunk 12]
```diff
@@ -34,12 +35,13 @@ public class Variable<T> {
     public func bind(to variable: Variable<T>) {
```
Comment (luispadron) [inline Snail/Variable.swift:34]: I think this function should instead return the subscription so we can get rid of the disposer. Callers of this function would then handle the subscription

File: Snail/Variable.swift [hunk 13]
```diff
@@ -34,12 +35,13 @@ public class Variable<T> {
...
         variable.asObservable().subscribe(onNext: { [weak self] value in
             self?.value = value
-        })
+        }).add(to: disposer)
     }
 
     public func map<U>(_ transform: @escaping (T) -> U) -> Variable<U> {
```
Comment (luispadron) [inline Snail/Variable.swift:40]: also think we can delete this function entirely since from what i've seen you should `asObservable` to use operators on variables.

File: SnailTests/ObservableTests.swift (hunk 11 above)
Comment (tim-chamberlin) [reply SnailTests/ObservableTests.swift:1037]: Yes, this test fails without the current changes and passes with them! woo! However, the other `Observable` operators internally initialize a subscriber that gets added to the observable's array of subscribers. Unfortunately I can't think of a good way to add this subscriber to the same disposer as any other subscribers of the observable..

**Review by luispadron — APPROVED** [review]
